File: dichotomy/dichotomy.py
from typing import Callable

class Dichotomy:
    def __init__(self, func: Callable[[float], float], a: float, b: float, eps: float):
        self.func = func
        self.a = a
        self.b = b
        self.eps = eps
    
    def gen(self) -> float:
        while (self.b - self.a) > self.eps:
            x = (self.a + self.b) / 2
            f1 = self.func(x - self.eps)
            f2 = self.func(x + self.eps)
            if f1 < f2:
                self.b = x
            else:
                self.a = x
        return self.b

def gen(func: Callable[[float], float], a: float, b: float, eps: float) -> float:
    return Dichotomy(func, a, b, eps)

File: lib/Dichotomy.java
import java.util.function.Function;

public class Dichotomy {
	private final Function<Double, Double> func;
	private double a, b;
	private final double eps;

	public Dichotomy(final Function<Double, Double> func, final double a, final double b, final double eps) {
		this.func = func;
		this.a = a;
		this.b = b;
		this.eps = eps;
	}

	public double gen() {
		while (Double.compare(b - a, eps) > 0) {
			final double x = (a + b) / 2;
			final double f1 = func.apply(x - eps);
			final double f2 = func.apply(x + eps);
			if (Double.compare(f1, f2) < 0) {
				b = x;
			} else {
				a = x;
			}
		}
		return b;
	}
}
